platform/darwin/src/MGLStyle.mm

```diff
@@ -96,21 +100,133 @@ + (NSURL *)emeraldStyleURL {
     return MGLStyleURL_emerald;
 }
 
-- (NSString *)name {
-    return @(self.mapView.mbglMap->getStyleName().c_str());
+#pragma mark -
+
+- (instancetype)initWithMapView:(MGLMapView *)mapView {
+    if (self = [super init]) {
+        _mapView = mapView;
+    }
+    return self;
 }
 
 - (NSURL *)URL {
     return [NSURL URLWithString:@(self.mapView.mbglMap->getStyleURL().c_str())];
 }
 
-- (MGLStyleLayer *)layerWithIdentifier:(NSString *)identifier
+- (NSString *)name {
+    return @(self.mapView.mbglMap->getStyleName().c_str());
+}
+
+#pragma mark Sources
+
+- (MGLSource *)sourceWithIdentifier:(NSString *)identifier
 {
-    auto mbglLayer = self.mapView.mbglMap->getLayer(identifier.UTF8String);
-    if (!mbglLayer) {
+    auto mbglSource = self.mapView.mbglMap->getSource(identifier.UTF8String);
+    if (!mbglSource) {
         return nil;
     }
 
+    // TODO: Fill in options specific to the respective source classes
+    // https://github.com/mapbox/mapbox-gl-native/issues/6584
+    MGLSource *source;
+    if (mbglSource->is<mbgl::style::VectorSource>()) {
+        source = [[MGLVectorSource alloc] initWithIdentifier:identifier];
+    } else if (mbglSource->is<mbgl::style::GeoJSONSource>()) {
+        source = [[MGLGeoJSONSource alloc] initWithIdentifier:identifier];
+    } else if (mbglSource->is<mbgl::style::RasterSource>()) {
+        source = [[MGLRasterSource alloc] initWithIdentifier:identifier];
+    } else {
+        source = [[MGLSource alloc] initWithIdentifier:identifier];
+    }
+
+    source.source = mbglSource;
+
+    return source;
+}
+
+- (void)addSource:(MGLSource *)source
+{
+    self.mapView.mbglMap->addSource(source.mbglSource);
+}
+
+- (void)removeSource:(MGLSource *)source
+{
+    self.mapView.mbglMap->removeSource(source.identifier.UTF8String);
+}
+
+#pragma mark Style layers
+
+- (NS_MUTABLE_ARRAY_OF(MGLStyleLayer *) *)layers
+{
+    auto layers = self.mapView.mbglMap->getLayers();
+    NSMutableArray *styleLayers = [NSMutableArray arrayWithCapacity:layers.size()];
+    for (auto layer = layers.rbegin(); layer != layers.rend(); ++layer) {
+        MGLStyleLayer *styleLayer = [self layerFromMBGLLayer:*layer];
+        [styleLayers addObject:styleLayer];
+    }
+    return styleLayers;
+}
+
+- (void)setLayers:(NS_MUTABLE_ARRAY_OF(MGLStyleLayer *) *)layers {
+    std::vector<std::unique_ptr<mbgl::style::Layer>> rawLayers;
+    rawLayers.reserve(layers.count);
+    for (MGLStyleLayer *layer in layers.reverseObjectEnumerator) {
+        rawLayers.push_back(std::unique_ptr<mbgl::style::Layer>(layer.layer));
```

Comment:
To respect ownership of the layers passed into `-setLayers:` – some of which may exist in the style and some of which may be new – we need to ability to remove a layer from the style and regain ownership of it in order to re-add it. So we’re blocked on `mbgl::Map::removeLayer()` returning ownership of the layer: #6959.